### PR TITLE
Add optional requirements for balance top listing

### LINKED ISSUE #5307: Exclude Baltop if player doesnt meet minimum balance // baltop-min-balance
### Feature description

A way to exclude players from /baltop if they dont meet a specific requirement

A few ways to do this could be:

- If they dont meet a minimum value
  - `baltop-requirement: 10000` *or `baltop-requirement: $10000`
  -  (If their balance isnt more than 10,000, they wouldnt show up in /baltop)
- A playtime based requirement 
  - `baltop-requirement: playtime` *or `baltop-requirement: 3600s`
    - `playtime: 3600` (1 hour in seconds)
  - (If they havent played for more than an hour, they wouldnt show up in /baltop)
- Balance isnt more than the starting balance set in the config.yml
  - `baltop-requirement: starting-balance`
  -  (If their balance value isnt more than  "starting-balance: 10,000", they wouldnt show up)
- Don't exclude
  - `baltop-requirement: none`
  - (Wouldnt exclude any players)

*Another way to do the minimum value &/ playtime is to just have the `baltop-requirement:` need either a `$` or `s`, so money requirement would be `baltop-requirement: $10000` and time requirement would be `baltop-requirement: 3600s`

Like mdcfe#1970 mentioned on the essx-discussion on Discord (message link: https://discord.com/channels/390942438061113344/390943808948535297/1095297860674785431 )

Food for thought:
There should be another setting under "`baltop-requirement`" (`show-negative: true/false`) where devs could make it so that negative balances show up or not, (If their balance is negative, it most likely means they're playing on the server)

### How the feature is useful

It would allow server developers to hide players who joined, got the starting balance or any kind of money, and then never joined the server. It would make /baltop to be less cramped and show players who actually play.

### PR DESCRIPTION
**Pull request by ayush03dev**
### Information

This PR closes #5307 

### Details

**Proposed feature:**  

This pull request adds two requirements that the player must meet in order for their name to appear in the balance top list. Those two requirements are: **1) Minimum Balance** and **2) Minimum Playtime**

The requirements can be configured from the following additions to config.yml file:

```
# Requirements which must be met by the player to get their name shown in the balance top list.
# Playtime is in seconds.
baltop-requirements:
  minimum-balance: 0
  minimum-playtime: 0
```

**Environments tested:**    

OS: Ubuntu 23.04

Java version:  17

- [x] Most recent Paper version (1.20.1, git-Paper-17)
- [ ] CraftBukkit/Spigot/Paper 1.12.2
- [ ] CraftBukkit 1.8.8

**Demonstration:**  

### Before configuring balance requirement (set to 0)
image
image

### After configuring balance requirement
image
image

As you can see the name disappears from the list because min-balance required was 500 but the player had 499.

### CODE REVIEW

**Review by JRoy — CHANGES_REQUESTED** [review]
You need to maintain backwards compatibility for the `show-zero-baltop` config option.

**Comment by ayush03dev** [comment]
> You need to maintain backwards compatibility for the `show-zero-baltop` config option.

I have re-added the above config option and made necessary checks along with the new options.

File: Essentials/src/main/java/com/earth2me/essentials/Settings.java [hunk 1]
```diff
@@ -1943,4 +1943,14 @@ public boolean isUpdateCheckEnabled() {
...
+    
+    @Override
+    public double getBaltopMinBalance() {
+        return config.getDouble("baltop-requirements.minimum-balance", 0);
+    }
+
+    @Override
+    public long getBaltopMinPlaytime() {
+        return config.getLong("baltop-requirements.minimum-playtime", 0);
+    }
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/Settings.java]: use config.getBigDecimal and return a BigDecimal

**Review by JRoy — CHANGES_REQUESTED** [review]

File: Essentials/src/main/java/com/earth2me/essentials/Settings.java (hunk 1 above)
Comment (ayush03dev) [reply Essentials/src/main/java/com/earth2me/essentials/Settings.java]: For only double right, the `getBaltopMinBalance` method?

Comment (JRoy) [reply Essentials/src/main/java/com/earth2me/essentials/Settings.java]: yeah :p

**Review by YanisBft — APPROVED** [review]

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandbalancetop.java [hunk 2]
```diff
@@ -112,14 +116,29 @@ public void run() {
...
+                            playtime = user.getBase().getStatistic(PLAY_ONE_TICK);
+                        }
+                        // Play time in seconds
+                        final long playTimeSecs = playtime / 20;
+
+                        // Checking if player meets the requirements of minimum balance and minimum playtime to be listed in baltop list
+                        if ((ess.getSettings().showZeroBaltop() || entry.getValue().getBalance().compareTo(BigDecimal.ZERO) > 0)
+                                && balance.compareTo(ess.getSettings().getBaltopMinBalance()) >= 0 &&
+                                playTimeSecs > ess.getSettings().getBaltopMinPlaytime()) {
                             newCache.getLines().add(tl("balanceTopLine", pos, entry.getValue().getDisplayName(), NumberUtil.displayCurrency(entry.getValue().getBalance(), ess)));
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandbalancetop.java]:
```suggestion
                        if ((ess.getSettings().showZeroBaltop() || balance.compareTo(BigDecimal.ZERO) > 0)
                                && balance.compareTo(ess.getSettings().getBaltopMinBalance()) >= 0 &&
                                playTimeSecs > ess.getSettings().getBaltopMinPlaytime()) {
                            newCache.getLines().add(tl("balanceTopLine", pos, entry.getValue().getDisplayName(), NumberUtil.displayCurrency(balance, ess)));
```

**Review by JRoy — APPROVED** [review]
Thanks for the contribution :)

**Pull request merged**